File: src/Submission_In_Python/Lab1/Question2/Question2.py
import collections

rows: int = int(input())
columns: int = int(input())
graph: [[str]] = []
cost: [[int]] = []

for i in range(rows):
    graph.append(input().split(" "))
    cost.append([int(1e9)] * columns)


def bfs(first_index: int, second_index: int):
    is_visited: [[bool]] = []
    for _i in range(rows):
        is_visited.append([False for _j in range(columns)])
    cur_cost: int = 1
    queue = collections.deque()
    queue.append((first_index, second_index))
    is_visited[first_index][second_index] = True
    # Time to check
    while len(queue) != 0:
        found_human: bool = False
        first: int = queue[0][0]
        second: int = queue[0][1]
        # where can I go now
        if first > 0 and not is_visited[first - 1][second] and graph[first - 1][second] == 'H':
            found_human = True
            queue.append((first - 1, second))
            cost[first - 1][second] = min(cost[first - 1][second], cur_cost)
            is_visited[first - 1][second] = True
        if first < rows - 1 and not is_visited[first + 1][second] and graph[first + 1][second] == 'H':
            found_human = True
            queue.append((first + 1, second))
            cost[first + 1][second] = min(cost[first + 1][second], cur_cost)
            is_visited[first + 1][second] = True
        if second > 0 and not is_visited[first][second - 1] and graph[first][second - 1] == 'H':
            found_human = True
            queue.append((first, second - 1))
            cost[first][second - 1] = min(cost[first][second - 1], cur_cost)
            is_visited[first][second - 1] = True
        if second < columns - 1 and not is_visited[first][second + 1] and graph[first][second + 1] == 'H':
            found_human = True
            queue.append((first, second + 1))
            cost[first][second + 1] = min(cost[first][second + 1], cur_cost)
            is_visited[first][second + 1] = True
        if found_human:
            cur_cost += 1
        queue.popleft()


for i in range(rows):
    for j in range(columns):
        if graph[i][j] == 'A':
            bfs(i, j)

min_time: int = int(-1e9)
number_of_survived_humans = 0
for i in range(rows):
    for j in range(columns):
        if graph[i][j] == 'H' and cost[i][j] != int(1e9):
            min_time = max(min_time, cost[i][j])
        if graph[i][j] == 'H' and cost[i][j] == int(1e9):
            number_of_survived_humans += 1

print(f"Time: {min_time} minutes")
print(f"No one survived" if number_of_survived_humans == 0 else f"{number_of_survived_humans} survived")

File: src/Submission_In_Java/Practice/Question2/Question2.java
package Submission_In_Java.Practice.Question2;

import java.io.*;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Question2 {
    private static final PrintWriter pw = new PrintWriter(new BufferedOutputStream(System.out), true);
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static char[][] graph;
    private static int[][] cost;
    private static int m, n;

    // Have not yet implemented the file input and output. Might want to do that when you can.
    public static void main(String[] args) throws IOException {
        m = Integer.parseInt(new StringTokenizer(br.readLine()).nextToken());
        n = Integer.parseInt(new StringTokenizer(br.readLine()).nextToken());
        graph = new char[m][];
        cost = new int[m][n];
        for (int i = 0; i < m; i++) {
            graph[i] = br.readLine().replace(" ", "").toCharArray();
            Arrays.fill(cost[i], Integer.MAX_VALUE);
        }
        br.close();
        // Time to print the things!!
        for (var s : graph) {
            pw.println(Arrays.toString(s));
        }
        // Time to do a BFS from each Alien Position
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (graph[i][j] == 'A') {
                    bfs(i, j);
                }
            }
        }

        for (var k : cost) {
            pw.println(Arrays.toString(k));
        }
        // Time to check -- nice!!
        int minTime = Integer.MIN_VALUE;
        int survived = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (graph[i][j] == 'H' && cost[i][j] != Integer.MAX_VALUE) {
                    minTime = Integer.max(minTime, cost[i][j]);
                }
                if (graph[i][j] == 'H' && cost[i][j] == Integer.MAX_VALUE) {
                    survived++;
                }
            }
        }
        pw.println("Min Time " + minTime);
        pw.println(survived == 0 ? "No one survived!" : survived + " survived");
        pw.close();
    }

    private static void bfs(int first, int second) {
        boolean[][] visited = new boolean[m][n];
        visited[first][second] = true;
        ArrayDeque<Pair> deque = new ArrayDeque<>();
        deque.addLast(new Pair(first, second));
        int curCost = 1;
        while (!deque.isEmpty()) {
            int i = deque.getFirst().first, j = deque.getFirst().second;
            boolean foundNewHuman = false;
            if (i > 0 && !visited[i - 1][j] && graph[i - 1][j] == 'H') {
                visited[i - 1][j] = true;
                cost[i - 1][j] = Integer.min(cost[i - 1][j], curCost);
                deque.addLast(new Pair(i - 1, j));
                foundNewHuman = true;
            }

            if (i < m - 1 && !visited[i + 1][j] && graph[i + 1][j] == 'H') {
                visited[i + 1][j] = true;
                cost[i + 1][j] = Integer.min(cost[i + 1][j], curCost);
                deque.addLast(new Pair(i + 1, j));
                foundNewHuman = true;
            }

            if (j > 0 && !visited[i][j - 1] && graph[i][j - 1] == 'H') {
                visited[i][j - 1] = true;
                cost[i][j - 1] = Integer.min(cost[i][j - 1], curCost);
                deque.addLast(new Pair(i, j - 1));
                foundNewHuman = true;
            }

            if (j < n - 1 && !visited[i][j + 1] && graph[i][j + 1] == 'H') {
                visited[i][j + 1] = true;
                cost[i][j + 1] = Integer.min(cost[i][j + 1], curCost);
                deque.addLast(new Pair(i, j + 1));
                foundNewHuman = true;
            }
            if (foundNewHuman) {
                curCost++;
            }
            deque.removeFirst();
        }
    }

    private record Pair(int first, int second) {

    }
}
